From 5d54b37562cb929029a3709bce700811366bc305 Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <bmastbergen@ciq.com>
Date: Thu, 1 May 2025 15:40:48 -0400
Subject: [PATCH] xen/blkfront: fix leaking data in shared pages
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

jira VULN-1437
cve CVE-2022-26365
commit-author Roger Pau Monne <roger.pau@citrix.com>
commit 2f446ffe9d737e9a844b97887919c4fda18246e7
upstream-diff One of the alloc_page calls has not been switched to
              GFP_KERNEL yet, so __GFP_ZERO is or'd with GFP_NOIO
              instead

When allocating pages to be used for shared communication with the
backend always zero them, this avoids leaking unintended data present
on the pages.

This is CVE-2022-26365, part of XSA-403.

	Signed-off-by: Roger Pau Monné <roger.pau@citrix.com>
	Reviewed-by: Jan Beulich <jbeulich@suse.com>
	Reviewed-by: Juergen Gross <jgross@suse.com>
	Signed-off-by: Juergen Gross <jgross@suse.com>
(cherry picked from commit 2f446ffe9d737e9a844b97887919c4fda18246e7)
	Signed-off-by: Brett Mastbergen <bmastbergen@ciq.com>
---
 drivers/block/xen-blkfront.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/block/xen-blkfront.c b/drivers/block/xen-blkfront.c
index 5cb25b62d3a8..f1223dc30bd2 100644
--- a/drivers/block/xen-blkfront.c
+++ b/drivers/block/xen-blkfront.c
@@ -201,7 +201,7 @@ static int fill_grant_buffer(struct blkfront_info *info, int num)
 			goto out_of_memory;
 
 		if (info->feature_persistent) {
-			granted_page = alloc_page(GFP_NOIO);
+			granted_page = alloc_page(GFP_NOIO | __GFP_ZERO);
 			if (!granted_page) {
 				kfree(gnt_list_entry);
 				goto out_of_memory;
@@ -1707,7 +1707,8 @@ static int blkfront_setup_indirect(struct blkfront_info *info)
 
 		BUG_ON(!list_empty(&info->indirect_pages));
 		for (i = 0; i < num; i++) {
-			struct page *indirect_page = alloc_page(GFP_NOIO);
+			struct page *indirect_page = alloc_page(GFP_NOIO |
+								__GFP_ZERO);
 			if (!indirect_page)
 				goto out_of_memory;
 			list_add(&indirect_page->lru, &info->indirect_pages);
-- 
2.43.5

